File: parallelization/R/fit.R
#' Fit a random forest model with sample `i` left out
#'
#' @param i Index of the observation to leave out
#' @param x Covariate data
#' @param y Response variable
#' @param ... Additional arguments to pass to `ranger::ranger()`
#'
#' @return Predicted class of the left out sample `i`
fit_rf_loo <- function(i, x, y, ...) {
  fit <- ranger::ranger(
    x = x[-i, , drop = FALSE],
    y = y[-i],
    num.threads = 1,
    ...
  )
  preds <- as.character(predict(fit, x[i, , drop = FALSE])$predictions)
  return(preds)
}


#' Fit a k-nearest neighbors model with sample `i` left out
#'
#' @param i Index of the observation to leave out
#' @param x Covariate data
#' @param y Response variable
#' @param ... Additional arguments to pass to `caret::train()`
#'
#' @return Predicted class of the left out sample `i`
fit_knn_loo <- function(i, x, y, ...) {
  train_control <- caret::trainControl(
    method = "cv",
    number = 5
  )
  # hyperparameter grid for k
  tune_grid <- expand.grid(k = c(5, 10, 15))
  fit <- caret::train(
    x = x[-i, , drop = FALSE],
    y = y[-i],
    method = "knn",
    ...,
    trControl = train_control,
    tuneGrid = tune_grid,
  )
  preds <- as.character(predict(fit, x[i, , drop = FALSE]))
  return(preds)
}

File: parallelization/python/fit.py
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
import numpy as np


def fit_rf_loo(i, x, y, **kwargs):
    """
    Fit a random forest model with sample `i` left out

    Parameters:
    -----------
    i: int
        The sample to leave out
    x: np.array
        Covariate matrix
    y: np.array
        Response vector
    **kwargs:
        Arbitrary keyword arguments for RandomForestClassifier

    Returns:
    --------
    str
        Predicted class of the left out sample `i`
    """
    # Remove the ith observation for leave-one-out
    x_train = np.delete(x, i, axis=0)
    y_train = np.delete(y, i)

    # Train the RandomForestClassifier
    rf = RandomForestClassifier(**kwargs)
    rf.fit(x_train, y_train)

    # Make a prediction on the left-out observation
    preds = rf.predict(x[i:(i+1), :])[0]
    return preds


def fit_knn_loo(i, x, y, **kwargs):
    """
    Fit a k-nearest neighbors model with sample `i` left out

    Parameters:
    -----------
    i: int
        The sample to leave out
    x: np.array
        Covariate matrix
    y: np.array
        Response vector
    **kwargs:
        Arbitrary keyword arguments for KNeighborsClassifier

    Returns:
    --------
    str
        Predicted class of the left out sample `i`
    """
    # Remove the ith observation for leave-one-out
    x_train = np.delete(x, i, axis=0)
    y_train = np.delete(y, i)

    # Train KNN and use grid search to find the best hyperparameters
    knn = KNeighborsClassifier(**kwargs)
    knn_grid_search = GridSearchCV(
        knn, 
        param_grid={
            "n_neighbors": [5, 10, 15]
        },
        cv=5
    )
    knn_grid_search.fit(x_train, y_train)
    best_knn = knn_grid_search.best_estimator_

    # Make a prediction on the left-out observation
    preds = best_knn.predict(x[i:(i+1), :])[0]
    return preds